### PR TITLE
Implement KIP-345 in aiokafka

### PR DESCRIPTION
**Pull request by patkivikram**
### Changes

Fixes #<https://github.com/aio-libs/aiokafka/issues/680>

Implements [KIP-345](https://cwiki.apache.org/confluence/display/KAFKA/KIP-345%3A+Introduce+static+membership+protocol+to+reduce+consumer+rebalances) in aiokafka

### Checklist

- [x] I think the code is well written
- [ ] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [ ] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` (e.g. `588.bugfix`)
  * if you don't have an `issue_id` change it to the pr id after creating the PR
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: `Fix issue with non-ascii contents in doctest text files.`

### CODE REVIEW

**Comment by patkivikram** [comment]
@tvoinarovskyi does this look okay?

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -267,7 +267,7 @@ jobs:
 
     - name: Install system dependencies
       run: |
-        sudo apt-get install -y libsnappy-dev libkrb5-dev krb5-user
+        sudo apt-get update -y libsnappy-dev libkrb5-dev krb5-user
```
Comment (forsberg) [inline .github/workflows/tests.yml:270]: sudo apt-get update && sudo apt-get install -y libsnappy-dev libkrb5-dev krb5-user

Comment (patkivikram) [reply .github/workflows/tests.yml]: still failiing ```Err:1 http://security.ubuntu.com/ubuntu bionic-updates/main amd64 libgssrpc4 amd64 1.16-2ubuntu0.1
  404  Not Found [IP: 52.147.219.192 80]
Err:2 http://security.ubuntu.com/ubuntu bionic-updates/main amd64 libkdb5-9 amd64 1.16-2ubuntu0.1
  404  Not Found [IP: 52.147.219.192 80]
Err:3 http://security.ubuntu.com/ubuntu bionic-updates/main amd64 libkadm5srv-mit11 amd64 1.16-2ubuntu0.1
  404  Not Found [IP: 52.147.219.192 80]
Err:4 http://security.ubuntu.com/ubuntu bionic-updates/main amd64 libkadm5clnt-mit11 amd64 1.16-2ubuntu0.1
  404  Not Found [IP: 52.147.219.192 80]
Err:6 http://security.ubuntu.com/ubuntu bionic-updates/main amd64 krb5-multidev amd64 1.16-2ubuntu0.1
  404  Not Found [IP: 52.147.219.192 80]
Err:7 http://security.ubuntu.com/ubuntu bionic-updates/main amd64 libkrb5-dev amd64 1.16-2ubuntu0.1
  404  Not Found [IP: 52.147.219.192 80]```

Comment (forsberg) [reply .github/workflows/tests.yml]: Try rebasing on master, there was a (very similar) fix there, and it builds successfully.

**Comment by reddyanilkumar** [comment]
unsubscribe
    On Wednesday, November 4, 2020, 04:18:33 AM GMT+5:30, Vikram Patki <notifications@github.com> wrote:  
 
 
Changes

Fixes ##680

Implements KIP-345 in aiokafka

Checklist
   
   -  I think the code is well written
   -  Unit tests for the changes exist
   -  Documentation reflects the changes
   -  Add a new news fragment into the CHANGES folder      
      - name it <issue_id>.<type> (e.g. 588.bugfix)
      - if you don't have an issue_id change it to the pr id after creating the PR
      - ensure type is one of the following:         
         - .feature: Signifying a new feature.
         - .bugfix: Signifying a bug fix.
         - .doc: Signifying a documentation improvement.
         - .removal: Signifying a deprecation or removal of public API.
         - .misc: A ticket has been closed, but it is not of interest to users.

      - Make sure to use full sentences with correct case and punctuation, for example: Fix issue with non-ascii contents in doctest text files.


You can view, comment on, or merge this pull request online at:

  https://github.com/aio-libs/aiokafka/pull/682

Commit Summary
   
   - Implement KIP-345 in aiokafka

File Changes
   
   -  M aiokafka/client.py (2) 
   -  A aiokafka/consumer/assignors.py (23) 
   -  M aiokafka/consumer/consumer.py (5) 
   -  M aiokafka/consumer/group_coordinator.py (158) 
   -  M aiokafka/errors.py (8) 
   -  A aiokafka/protocol/group.py (75) 

Patch Links:
   
   - https://github.com/aio-libs/aiokafka/pull/682.patch
   - https://github.com/aio-libs/aiokafka/pull/682.diff

File: aiokafka/client.py [hunk 2]
```diff
@@ -574,6 +575,7 @@ def _check_api_version_response(self, response):
         # in descending order. As soon as we find one that works, return it
         test_cases = [
             # format (<broker version>, <needed struct>)
+            ((2, 3, 0), SyncGroupRequest[0].API_KEY, 3),
```
Comment (tvoinarovskyi) [inline aiokafka/client.py:578]: Those checks are meant to pick a working version, why would you introduce a 2nd check for version 2.3.0? What version introduces SyncGroup v3?

**Comment by tvoinarovskyi** [comment]
Hi @patkivikram, looks great, did you use some linter for the files here? It's really hard to review because of the big amount of changes unrelated to logic. Would it be hard to either do a separate PR if the linter used is "black" or revert the lint changes? Sorry for the trouble.

File: aiokafka/client.py (hunk 2 above)
Comment (patkivikram) [reply aiokafka/client.py:578]: It is introduced in version 2.3

**Comment by patkivikram** [comment]
@tvoinarovskyi let me know if this looks good

File: aiokafka/consumer/assignors.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import abc
+
+
+class AbstractStaticPartitionAssignor(AbstractPartitionAssignor):
+    """
+    Abstract assignor implementation that also supports static assignments (KIP-345)
+    """
+
+    @abc.abstractmethod
+    def assign(self, cluster, members, member_group_instance_ids):
```
Comment (forsberg) [inline aiokafka/consumer/assignors.py:11]: `def assign(self, cluster, members, member_group_instance_ids): # lgtm[py/inheritance/signature-mismatch]` 

..to inform LGTM that you do really mean to have a signature mismatch.

https://lgtm.com/help/lgtm/alert-suppression#python

**Comment by deed02392** [comment]
Hi all, just wondering if this PR has an ETA?

**Comment by forsberg** [comment]
What's left to do to get this merged? Can I help? @tvoinarovskyi - any insights?

**Comment by tvoinarovskyi** [comment]
Hi, thanks for the work on this, now it's rally clear and I would love to have this merged. The only issue I have is the abscense of tests and I am not familiar with the Static group management enough. Could you share how did you test the change, can you try preparing a test to reproduce the membership change? Or I could try to do it if I know how to reproduce it manually.

**Comment by g-clef** [comment]
Here's how I reproduce it manually (which I'm not sure if you can automate...it's a little messy):

- Assume a topic with multiple partitions and multiple consumers.
- In non-KIP-345-mode
  - When one consumer leaves, the other consumers immediately rebalance.
  - When the original consumer returns, another rebalance occurs and the original consumer gets partitions that may or may not include its original assignments, depending on the negotiation in the rebalance.
- In KIP-345-mode
  - When one consumer leaves, no rebalance occurs until after that consumer's `session timeout` expires.
  - If the original consumer rejoins the group before its `session timeout` expiration, the consumer is assigned the same partitions it had before leaving.

I've test this manually by running consumers locally and watching what happens when I kill or stop one of them, which I acknowledge is not the greatest way to test something.

File: aiokafka/consumer/group_coordinator.py [hunk 4]
```diff
@@ -402,8 +410,18 @@ async def _perform_assignment(
             'Invalid assignment protocol: %s' % assignment_strategy
         member_metadata = {}
         all_subscribed_topics = set()
-        for member_id, metadata_bytes in members:
+        group_instance_id_mapping = {}
```
Comment (PhilipGarnero) [inline aiokafka/consumer/group_coordinator.py:413]: Not sure what this is used for as it gets replaced by `group_instance_id` a few lines down and that it only serves for AbstractStaticPartitionAssignor. Couldn't we directly use `group_instance_id` instead ?

File: aiokafka/consumer/group_coordinator.py [hunk 5]
```diff
@@ -421,7 +439,12 @@ async def _perform_assignment(
                   " with subscriptions %s", self.group_id, assignor.name,
                   member_metadata)
 
-        assignments = assignor.assign(self._cluster, member_metadata)
+        if isinstance(assignor, AbstractStaticPartitionAssignor):
```
Comment (PhilipGarnero) [inline aiokafka/consumer/group_coordinator.py:442]: Maybe we could put a warning if `group_instance_id` is not None and no `AbstractStaticPartitionAssignor` is found.

**Comment by PhilipGarnero** [comment]
This work seems pretty good to me. What is missing for the merge ?

**Comment by oferziss-armis** [comment]
hi guys when can we expect this to be released? seems that it's ready what is the ETL here?

**Comment by ntextreme3** [comment]
Is there anything that still needs to be done here?

**Comment by ods** [comment]
> Is there anything that still needs to be done here?

We need to rebase it (resolve conflicts) and write tests at least.

**Comment by ntextreme3** [comment]
@ods I rebased this in https://github.com/ntextreme3/aiokafka-1 -- any suggestions on the best way to get that here?

**Comment by ods** [comment]
> any suggestions on the best way to get that here?

Not sure about the best way, but creating a new PR should work.

**Comment by wbarnha** [comment]
This PR is definitely useful but I believe this should be closed and resumed in #827.

**Pull request closed without merging**